File: lab/python/textos/contagem_de_caracteres_especificos.py
import re


def contar_caracteres_especificos(string, caracter):
    if not caracter or len(caracter) != 1:
        raise ValueError(
            "O argumento 'caracter' deve conter apenas um caractere."
        )
    string = string.lower()
    caracter = caracter.lower()
    ocorrencias = 0
    for palavra in string:
        if palavra == caracter:
            ocorrencias += 1

    return ocorrencias


def contar_caracteres_especificos2(string, caracter):
    return string.lower().count(caracter.lower())


def contar_caracteres_especificos3(string, caracter):
    # rf = f-string para regex
    padrao = rf"[{caracter}áàâãäÁÀÂÃÄ]"
    return len(re.findall(padrao, string))


def contar_caracteres_especificos4(string, caracter):
    return sum(1 for c in string.lower() if c == caracter.lower())


print(contar_caracteres_especificos("abacaxi", 'a'))
print(contar_caracteres_especificos2("abacaxi", 'a'))
print(contar_caracteres_especificos3("abacaxi", 'a'))
print(contar_caracteres_especificos4("abacaxi", 'a'))

File: lab/javascript/textos/contagem_de_caracteres_especificos.js
function contarCaracteres1(string, caracter) {
  if (!caracter || caracter.length !== 1) {
    throw new Error("O argumento 'caracter' deve conter apenas um caractere.");
  }
  string = string.toLowerCase();
  caracter = caracter.toLowerCase();
  let ocorrencias = 0;

  for (const c of string) {
    if (c === caracter) ocorrencias++;
  }
  return ocorrencias;
}

function contarCaracteres2(string, caracter) {
  if (!caracter || caracter.length !== 1)
    throw new Error("O argumento 'caracter' deve conter apenas um caractere.");
  return string.toLowerCase().split(caracter.toLowerCase()).length - 1;
}

function contarCaracteres3(string, caracter) {
  // Regex com acentos
  const padrao = new RegExp(`[${caracter}áàâãäÁÀÂÃÄ]`, "g");
  const encontrados = string.match(padrao);
  return encontrados ? encontrados.length : 0;
}

function contarCaracteres4(string, caracter) {
  let cont = 0;
  const alvo = caracter.toLowerCase();
  for (const c of string.toLowerCase()) {
    if (c === alvo) cont++;
  }
  return cont;
}

// ---- Testes ----
console.log(contarCaracteres1("abacaxi", "a"));
console.log(contarCaracteres2("abacaxi", "a"));
console.log(contarCaracteres3("abacaxi", "a"));
console.log(contarCaracteres4("abacaxi", "a"));
